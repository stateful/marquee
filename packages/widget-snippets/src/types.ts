import type { ContextProperties } from "@vscode-marquee/utils";
import type { vscLanguages } from './constants';
import SnippetModel from './models/Snippet';

export type VSCLanguages = keyof typeof vscLanguages;
export type Snippet = SnippetModel;

export interface SnippetTreeItem {
  item: Snippet
  isTreeItem: boolean
}

export interface Selection {
  label: string
  snippet: Snippet
}

export interface State {
  snippets: Snippet[]
  snippetFilter: string
  snippetSelected: string
  snippetSplitter?: number
}

export interface Events {
  openSnippet: string
  selectSnippet: string
  addSnippet: Snippet
  addNote: Pick<Snippet, keyof Snippet>
}

export interface Context extends ContextProperties<State> {
  _addSnippet: (
    snippet: Pick<Snippet, 'title' | 'body'>,
    isWorkspaceTodo: boolean
  ) => string
  _removeSnippet: (id: string) => void
  _updateSnippet: (snippet: any) => void
}
