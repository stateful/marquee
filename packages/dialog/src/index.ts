import DialogContainer from './DialogContainer'
import DialogTitle from './DialogTitle'

export { DialogContainer, DialogTitle }
