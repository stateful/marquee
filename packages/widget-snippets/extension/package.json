{
  "name": "@vscode-marquee/widget-snippets-extension",
  "description": "Extension logic for Snippets Widget",
  "main": "../build/extension.js",
  "module": "../build/extension.js",
  "types": "../build/extension.d.ts"
}
