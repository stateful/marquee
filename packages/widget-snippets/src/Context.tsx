import React, { createContext, useState, useEffect } from "react";
import { connect, getEventListener, MarqueeWindow } from "@vscode-marquee/utils";

import AddDialog from "./dialogs/AddDialog";
import EditDialog from "./dialogs/EditDialog";
import type { State, Context, Snippet, Events } from './types';

declare const window: MarqueeWindow;
const SnippetContext = createContext<Context>({} as Context);
const WIDGET_ID = '@vscode-marquee/snippets-widget';

const SnippetProvider = ({ children }: { children: React.ReactElement }) => {
  const eventListener = getEventListener<Events>();
  const widgetState = getEventListener<State>(WIDGET_ID);
  const providerValues = connect<State>(window.marqueeStateConfiguration[WIDGET_ID].state, widgetState);

  const [showAddDialog, setShowAddDialog] = useState(false);
  const [showEditDialog, setShowEditDialog] = useState<string | undefined>();

  const _addSnippet = (
    snippet: Pick<Snippet, 'title' | 'body' | 'language'>,
    isWorkspaceTodo: boolean
  ): string => {
    const globalSnippets = providerValues.snippets;
    const id = [...Array(8)].map(() => Math.random().toString(36)[2]).join('');

    const newSnippet: Snippet = Object.assign({}, snippet, {
      id,
      archived: false,
      createdAt: new Date().getTime(),
      workspaceId: isWorkspaceTodo
        ? window.activeWorkspace?.id || null
        : null
    });

    globalSnippets.unshift(newSnippet);
    providerValues.setSnippets(globalSnippets);
    return id;
  };

  useEffect(() => {
    eventListener.on('openAddSnippetDialog', setShowAddDialog);
    eventListener.on('openEditSnippetDialog', setShowEditDialog);
    eventListener.on('addSnippet', (snippet) => _addSnippet(
      snippet,
      snippet.workspaceId === window.activeWorkspace?.id
    ));

    return () => {
      widgetState.removeAllListeners();
      eventListener.removeAllListeners();
    };
  }, []);

  const _removeSnippet = (id: string) => {
    const globalSnippets = providerValues.snippets;
    const index = globalSnippets.findIndex((snippet) => snippet.id === id);

    if (index < 0) {
      return console.error(`Couldn't find note with id "${id}"`);
    }

    globalSnippets.splice(index, 1);
    providerValues.setSnippets(globalSnippets);
  };

  const _updateSnippet = (snippet: Snippet) => {
    const globalSnippets = providerValues.snippets;
    const index = globalSnippets.findIndex((s) => s.id === snippet.id);

    if (index < 0) {
      return console.error(`Couldn't find note with id "${snippet.id}"`);
    }

    globalSnippets[index] = snippet;
    providerValues.setSnippets(globalSnippets);
  };

  return (
    <SnippetContext.Provider
      value={{
        ...providerValues,
        _addSnippet,
        _removeSnippet,
        _updateSnippet,

        showAddDialog,
        setShowAddDialog,
        showEditDialog,
        setShowEditDialog
      }}
    >
      { showAddDialog && (
        <AddDialog close={() => setShowAddDialog(false)} />
      )}
      { showEditDialog && providerValues.snippets.find((s) => showEditDialog === s.id) && (
        <EditDialog snippet={providerValues.snippets.find((s) => showEditDialog === s.id)!} close={() => setShowEditDialog(undefined) } />
      )}
      {children}
    </SnippetContext.Provider>
  );
};

export default SnippetContext;

export { SnippetProvider };
