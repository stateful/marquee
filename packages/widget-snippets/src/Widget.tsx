import React, { useContext, useEffect, useMemo, useCallback } from "react";
import { makeStyles } from "@material-ui/core/styles";
import {
  Grid,
  IconButton,
  Typography,
  TextField,
  Button,
} from "@material-ui/core";

import { AddCircle, Clear } from "@material-ui/icons";
import LinkIcon from "@material-ui/icons/Link";
import wrapper, { Dragger, HidePop } from "@vscode-marquee/widget";
import { GlobalContext, jumpTo, DoubleClickHelper, MarqueeWindow } from "@vscode-marquee/utils";

import SplitterLayout from "react-splitter-layout";
import { List, AutoSizer } from "react-virtualized";
import "react-virtualized/styles.css";
import "../css/react-splitter-layout.css";

import SnippetContext, { SnippetProvider } from "./Context";

import SnippetEditor from "./components/Editor";
import SnippetListItem from "./components/ListItem";

declare const window: MarqueeWindow;

const useStyles = makeStyles(() => ({
  root: {
    width: "100%",
    height: "100%",
    maxWidth: "100%",
  },
  widgetTitle: {
    borderBottom: "1px solid var(--vscode-foreground)",
    padding: "8px",
  },
}));

interface RowRendererProps {
  key: React.Key
  index: number
  style: object
}

let Snippets = () => {
  const classes = useStyles();
  const { globalScope } = useContext(GlobalContext);

  const {
    _updateSnippet,
    setSnippetFilter,
    setSnippetSelected,
    setSnippetSplitter,
    snippets,
    snippetFilter,
    snippetSelected,
    snippetSplitter,

    setShowAddDialog,
    setShowEditDialog
  } = useContext(SnippetContext);

  const snippet = useMemo(() => {
    return snippets.find((snippet) => snippet.id === snippetSelected);
  }, [snippetSelected, snippets]);

  const snippetLinkFileName = useMemo(() => {
    if (
      snippet && snippet.path && snippet.exists && snippet.hasOwnProperty("path") &&
      snippet.path.indexOf("/") !== -1
    ) {
      return snippet.path.split("/").reverse()[0].toUpperCase();
    }

    return '';
  }, [snippet]);

  const snippetsArr = useMemo(() => {
    let filteredItems = snippets;

    if (!globalScope) {
      filteredItems = filteredItems.filter((item) => (
        item.workspaceId && item.workspaceId === window.activeWorkspace?.id
      ));
    }

    if (snippetFilter) {
      filteredItems = filteredItems.filter((item) => (
        item.title.toLowerCase().indexOf(snippetFilter.toLowerCase()) !== -1
      ));
    }

    return filteredItems;
  }, [globalScope, snippets, snippetFilter]);

  useEffect(() => {
    if (snippetsArr.length !== 0) {
      setSnippetSelected(snippetsArr[0].id);
    }
  }, [snippetFilter, globalScope]);

  const snippetItemClick = useCallback((e, index) => {
    if (e.detail === 1) {
      if (snippetsArr[index] && snippetsArr[index].id) {
        setSnippetSelected(snippetsArr[index].id);
      } else {
        console.error({
          message:
            "Trying to set selected snippet to an entry that doesn't exist in the array",
          eventObj: e,
        });
      }
    }

    if (e.detail === 2) {
      setShowEditDialog(snippetsArr[index].id);
    }
  }, [snippetsArr]);

  const rowRenderer = ({ key, index, style }: RowRendererProps) => {
    const snippetEntry = snippetsArr[index];
    return (
      <SnippetListItem
        key={key}
        keyVal={key}
        index={index}
        style={style}
        snippet={snippetEntry}
        selected={snippetSelected}
        click={snippetItemClick}
      />
    );
  };

  return (
    <>
      <Grid item style={{ maxWidth: "100%" }}>
        <div className={classes.widgetTitle}>
          <Grid container direction="row" justifyContent="space-between">
            <Grid item>
              <Grid container direction="row" spacing={1} alignItems="center">
                <Grid item>
                  <Typography variant="subtitle1">Snippets</Typography>
                </Grid>
                <Grid item>
                  {snippet &&
                    snippetLinkFileName !== "" &&
                    snippetsArr.length !== 0 && (
                      <Button
                        size="small"
                        startIcon={<LinkIcon />}
                        disableFocusRipple
                        onClick={() => jumpTo(snippet)}
                      >
                        {snippetLinkFileName}
                      </Button>
                    )}
                </Grid>
              </Grid>
            </Grid>
            <Grid item>
              <Grid container direction="row" spacing={1} alignItems="center">
                <Grid item>
                  <IconButton
                    size="small"
                    onClick={() => setShowAddDialog(true)}
                  >
                    <AddCircle fontSize="small" />
                  </IconButton>
                </Grid>
                <Grid item>
                  <DoubleClickHelper content="Double-click a snippet title to edit and right-click for copy & paste" />
                </Grid>
                <Grid item>
                  <HidePop name="snippets" />
                </Grid>
                <Grid item>
                  <Dragger />
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </div>
      </Grid>
      <Grid item xs>
        <Grid
          container
          wrap="nowrap"
          direction="column"
          style={{ height: "100%" }}
        >
          <Grid item xs style={{ overflow: "hidden" }}>
            <SplitterLayout
              percentage={true}
              primaryIndex={0}
              secondaryMinSize={10}
              primaryMinSize={10}
              secondaryInitialSize={snippetSplitter || 80}
              onSecondaryPaneSizeChange={setSnippetSplitter}
            >
              <div
                style={{
                  height: "100%",
                  overflow: "hidden",
                }}
              >
                <Grid
                  container
                  wrap="nowrap"
                  direction="column"
                  style={{
                    height: "100%",
                    overflow: "hidden",
                  }}
                >
                  <Grid item style={{ maxWidth: "100%", padding: "8px" }}>
                    <TextField
                      margin="dense"
                      placeholder="Filter..."
                      fullWidth
                      size="small"
                      value={snippetFilter}
                      onChange={(e) => setSnippetFilter(e.target.value)}
                      InputProps={{
                        endAdornment: (
                          <Clear
                            fontSize="small"
                            style={{ cursor: "pointer" }}
                            onClick={() => setSnippetFilter("") }
                          />
                        ),
                      }}
                    />
                  </Grid>
                  <Grid item xs style={{ maxWidth: "100%" }}>
                    <AutoSizer>
                      {({ height, width }) => (
                        <List
                          width={width}
                          height={height}
                          rowCount={snippetsArr.length}
                          rowHeight={30}
                          rowRenderer={rowRenderer}
                        />
                      )}
                    </AutoSizer>
                  </Grid>
                </Grid>
              </div>
              <div style={{ height: "100%" }}>
                <Grid container style={{ width: "100%", height: "100%" }}>
                  <Grid
                    item
                    style={{
                      height: "100%",
                      width: "100%",
                      overflow: "auto",
                    }}
                  >
                    {snippetsArr.length === 0 && (
                      <Grid
                        container
                        style={{ height: "100%" }}
                        alignItems="center"
                        justifyContent="center"
                        direction="column"
                      >
                        <Grid item>
                          <Typography>Nothing here yet.</Typography>
                        </Grid>
                        <Grid item>&nbsp;</Grid>
                        <Grid item>
                          <Button
                            startIcon={<AddCircle />}
                            variant="outlined"
                            onClick={() => setShowAddDialog(true) }
                          >
                            Create a snippet
                          </Button>
                        </Grid>
                      </Grid>
                    )}
                    {snippetsArr.length !== 0 && snippet && (
                      <SnippetEditor
                        body={snippet.body}
                        language={snippet.language}
                        onChange={(body) => _updateSnippet({ ...snippet, body }) }
                      />
                    )}
                  </Grid>
                </Grid>
              </div>
            </SplitterLayout>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

export default wrapper(() => (
  <SnippetProvider>
    <Snippets />
  </SnippetProvider>
), 'snippets');
