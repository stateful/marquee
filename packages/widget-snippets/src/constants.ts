import type { State } from './types';

export const DEFAULT_STATE: State = {
  snippets: [],
  snippetFilter: '',
  snippetSelected: '',
  snippetSplitter: undefined
};

export const languages = [
  { name: "python", value: "python" },
  { name: "markdown", value: "markdown" },
  { name: "clike", value: "clike" },
  { name: "shell", value: "shell" },
  { name: "css", value: "css" },
  { name: "sass", value: "sass" },
  { name: "scss", value: "sass" },
  { name: "sql", value: "sql" },
  { name: "stylus", value: "stylus" },
  { name: "tsx", value: "jsx" },
  { name: "javascript", value: "javascript" },
  { name: "diff", value: "diff" },
  { name: "git", value: "shell" },
  { name: "go", value: "go" },
  { name: "toml", value: "toml" },
  { name: "json", value: "javascript" },
  { name: "less", value: "css" },
  { name: "cmake", value: "cmake" },
  { name: "makefile", value: "shell" },
  { name: "objectivec", value: "clike" },
  { name: "yaml", value: "yaml" },
  { name: "markup", value: "htmlmixed" },
  { name: "jsx", value: "jsx" },
  { name: "text", value: "text" },
  { name: "xml", value: "xml" },
  { name: "vue", value: "vue" },
  { name: "rust", value: "rust" },
  { name: "ruby", value: "ruby" },
  { name: "html", value: "htmlmixed" },
] as const;

export const vscLanguages = {
  markdown: "markdown",
  text: "text",
  yaml: "yaml",
  sql: "sql",
  typescriptreact: "jsx",
  typescript: "javascript",
  scss: "scss",
  sass: "sass",
  css: "css",
  less: "less",
  rust: "rust",
  ruby: "clike",
  python: "python",
  php: "clike",
  cpp: "c",
  diff: "diff",
  "git-commit": "git",
  "git-rebase": "git",
  html: "markup",
  shellscript: "shell",
  javascriptreact: "jsx",
  javascript: "javascript",
  json: "json",
  makefile: "shell",
  plaintext: "text",
  xml: "markup",
  jsonc: "json",
  "objective-c": "objectivec",
  go: "go",
} as const;
