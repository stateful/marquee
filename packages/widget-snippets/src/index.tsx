import React from 'react';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCode } from "@fortawesome/free-solid-svg-icons";

import Snippets from "./Widget";

export default {
  name: "snippets",
  icon: <FontAwesomeIcon icon={faCode} />,
  tags: ["productivity", "workflow", "search", "organize"],
  description:
    "Create or extract code snippets then edit, organize and insert them directly into code.",
  component: Snippets,
};
export { Snippets };
