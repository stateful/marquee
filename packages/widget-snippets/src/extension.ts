import vscode from 'vscode'
import type { Client } from 'tangle'

import ExtensionManager from '@vscode-marquee/utils/extension'

import Snippet from './models/Snippet'
import ContentProvider from './provider/ContentProvider'
import SnippetStorageProvider from './provider/SnippetStorageProvider'
import { DEFAULT_STATE, STATE_KEY } from './constants'
import type { SnippetTreeItem, State, Events, Selection } from './types'

export class SnippetExtensionManager extends ExtensionManager<State, {}> {
  private _contentProvider = new ContentProvider()
  private _fsProvider: SnippetStorageProvider

  constructor (context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
    super(context, channel, STATE_KEY, {}, DEFAULT_STATE)
    this._fsProvider = new SnippetStorageProvider(context, channel, this.getActiveWorkspace()?.id)

    this._disposables.push(
      vscode.commands.registerCommand('marquee.snippet.move', this._moveSnippet.bind(this)),
      vscode.commands.registerCommand('marquee.snippet.addEmpty', this._addEmptySnippet.bind(this)),
      vscode.commands.registerCommand('marquee.snippet.insert', this._insertFromTreeView.bind(this)),
      vscode.commands.registerCommand('marquee.snippet.delete', this._removeSnippet.bind(this)),
      vscode.commands.registerCommand('marquee.snippet.copyToClipboard', this._copyToClipboard.bind(this)),

      vscode.commands.registerTextEditorCommand('marquee.snippet.insertEditor', this._insertEditor.bind(this)),
      vscode.commands.registerTextEditorCommand('marquee.snippet.addEditor', this._addSnippet.bind(this)),
      vscode.workspace.registerTextDocumentContentProvider(ContentProvider.scheme, this._contentProvider),
      vscode.workspace.registerFileSystemProvider(
        SnippetStorageProvider.scheme,
        this._fsProvider,
        { isCaseSensitive: true }
      )
    )

    this._fsProvider.on('saveNewSnippet', this._saveNewSnippet.bind(this))
    this._fsProvider.on('updateSnippet', this._updateSnippet.bind(this))
  }

  private async _saveNewSnippet (snippet: Snippet) {
    const newSnippets = [snippet].concat(this.state.snippets)
    this.updateState('snippets', newSnippets)
    this.broadcast({ snippets: newSnippets })

    this._openSnippet(snippet.path!)
    await vscode.commands.executeCommand('workbench.action.closeActiveEditor')
    vscode.commands.executeCommand('marquee.refreshCodeActions')
    vscode.window.showInformationMessage(
      `Added ${snippet.title} to your snippets in Marquee`,
      'Open Marquee'
    ).then((item) => item && this.emit('gui.open'))
  }

  private _updateSnippet (snippet: Snippet) {
    const index = this.state.snippets.findIndex((s) => s.id === snippet.id)
    if (index < 0) {
      return vscode.window.showErrorMessage(`Couldn't update snippet "${snippet.title}"`)
    }

    this.state.snippets[index].body = snippet.body
    this.updateState('snippets', this.state.snippets)
    this.broadcast({ snippets: this.state.snippets })
  }

  private async _addEmptySnippet () {
    return this._openSnippet('/New Snippet')
  }

  private async _openSnippet (path: string) {
    const setting: vscode.Uri = vscode.Uri.parse(`${SnippetStorageProvider.scheme}:${path}`)
    const doc = await vscode.workspace.openTextDocument(setting)
    await vscode.window.showTextDocument(doc, 2, false)
  }

  /**
   * add snippet into text editor
   */
  private _addSnippet (editor: vscode.TextEditor) {
    const { path, text, name } = this.getTextSelection(editor)

    if (text.length < 1) {
      return vscode.window.showWarningMessage('Marquee: no text selected')
    }

    const id = this.generateId()
    const snippet = new Snippet(
      this.getActiveWorkspace()?.id || null,
      name,
      text,
      false,
      Date.now(),
      id,
      path
    )
    const newSnippets = [snippet].concat(this.state.snippets)
    this.updateState('snippets', newSnippets)
    this.broadcast({ snippets: newSnippets })

    vscode.commands.executeCommand('marquee.refreshCodeActions')
    vscode.window.showInformationMessage(
      `Added ${name} to your snippets in Marquee`,
      'Open Marquee'
    ).then((item) => item && this.emit('gui.open'))
  }

  /**
   * insert snippet
   * this action is triggerd by either clicking on a snippet tree view item
   * or clicking on "Insert" in the context menu
   */
  private _insertFromTreeView (item: Snippet | SnippetTreeItem) {
    const snippet = (item as SnippetTreeItem).isTreeItem
      ? (item as SnippetTreeItem).item
      : item as Snippet
    const activeTextEditor = vscode.window.activeTextEditor

    if (!activeTextEditor) {
      return vscode.window.showInformationMessage(
        'Marquee: No editor open, a snippet can only be inserted into a text document.'
      )
    }

    activeTextEditor.edit((editor) => {
      if (!activeTextEditor.selection) {
        return console.warn('No text selected')
      }

      const pos = activeTextEditor.selection
      editor.insert(pos.active, snippet.body)
      activeTextEditor.revealRange(
        new vscode.Range(pos.start, pos.end),
        vscode.TextEditorRevealType.InCenterIfOutsideViewport
      )
      vscode.window.showTextDocument(activeTextEditor.document)
    })
  }

  private _copyToClipboard ({ item }: SnippetTreeItem) {
    vscode.env.clipboard.writeText(item.body)
  }

  /**
   * remove snippet after clicking on "Remove" through the context menu
   * in the tree view
   */
  private _removeSnippet ({ item }: SnippetTreeItem) {
    const newSnippets = this.state.snippets.filter((s) => s.id !== item.id)
    this.updateState('snippets', newSnippets)
    this.broadcast({ snippets: newSnippets })
  }

  /**
   * insert snippet into editor
   */
  private async _insertEditor (editor: vscode.TextEditor) {
    const snippet = await this._pickSnippet()
    editor.insertSnippet(
      new vscode.SnippetString(snippet.body),
      editor.selection.start
    )
  }

  private _pickSnippet () {
    return new Promise<Snippet>((resolve) => {
      const qp = vscode.window.createQuickPick<Selection>()
      qp.items = this.state.snippets.map((snippet: Snippet): Selection => ({
        label: snippet.title,
        snippet
      }))
      qp.onDidChangeSelection((selection) => {
        resolve(selection[0].snippet)
        qp.hide()
      })
      qp.onDidHide(() => qp.dispose())
      qp.show()
    })
  }

  /**
   * archive todo
   * @param item TreeView item that represents a todo in a tree view
   * @returns (un)archived todo
   */
  private async _moveSnippet (item: { id: string, archived: boolean }) {
    if (!item) {
      item = await this._pickSnippet()
    }

    const awsp = this.getActiveWorkspace()

    if (!awsp) {
      return console.warn('Can\'t move snippet, no workspace selected')
    }

    const snippetIndex = this.state.snippets.findIndex((s) => s.id === item.id)

    if (snippetIndex < 0) {
      return console.warn(`Couldn't find snippet to toggle with id "${item.id}"`)
    }

    this.state.snippets[snippetIndex].workspaceId = awsp.id
    this.updateState('snippets', this.state.snippets)
  }

  public setBroadcaster (tangle: Client<State>) {
    super.setBroadcaster(tangle)
    const _tangle = this._tangle as any as Client<Events>

    /**
     * open snippet document by webview request
     */
    _tangle.on('openSnippet', (path: string) => this._openSnippet(path))

    /**
     * select snippet in webview when file is opened
     */
    this._fsProvider.on('fileStat', (id: string) => _tangle.emit('selectSnippet', id))

    return this
  }
}

export function activate (
  context: vscode.ExtensionContext,
  channel: vscode.OutputChannel
) {
  const stateManager = new SnippetExtensionManager(context, channel)

  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  }
}
export * from './types'
