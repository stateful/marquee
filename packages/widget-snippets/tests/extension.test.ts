import path from 'path'
import vscode from 'vscode'
import { SnippetExtensionManager, activate } from '../src/extension'

jest.mock('../src/provider/SnippetStorageProvider', () => (class {
  on = jest.fn()
}))

jest.useFakeTimers().setSystemTime(new Date('2020-01-01').getTime())

beforeEach(() => {
  (vscode.window.showInformationMessage as jest.Mock).mockClear();
  (vscode.window.showTextDocument as jest.Mock).mockClear()
})

test('should register all commands in constructor', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  expect(m['_fsProvider'].on).toBeCalledTimes(2)
  expect(vscode.commands.registerCommand).toBeCalledTimes(5)
  expect(vscode.commands.registerTextEditorCommand).toBeCalledTimes(2)
  expect(vscode.workspace.registerTextDocumentContentProvider).toBeCalledTimes(1)
  expect(vscode.workspace.registerFileSystemProvider).toBeCalledTimes(1)
})

test('_saveNewSnippet', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  m['_openSnippet'] = jest.fn()
  m.state.snippets = [{ foo: 'bar' } as any]
  m['_saveNewSnippet']({ bar: 'foo', path: 'foobar' } as any)
  expect((m.updateState as jest.Mock).mock.calls).toMatchSnapshot()
  expect((m['broadcast'] as jest.Mock).mock.calls).toMatchSnapshot()
  expect(m['_openSnippet']).toBeCalledWith('foobar')
})

test('_updateSnippet', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  m.state.snippets = [{ id: 'foo' } as any]
  m['_updateSnippet']({ id: 'bar' } as any)
  expect(vscode.window.showErrorMessage).toBeCalledTimes(1)
  m['_updateSnippet']({ id: 'foo', body: 'bar' } as any)
  expect((m.updateState as jest.Mock).mock.calls).toMatchSnapshot()
  expect((m['broadcast'] as jest.Mock).mock.calls).toMatchSnapshot()
})

test('_addEmptySnippet', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  m['_openSnippet'] = jest.fn()
  m['_addEmptySnippet']()
  expect(m['_openSnippet']).toBeCalledWith('/New Clipboard Item')
})

test('_openSnippet', async () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  await m['_openSnippet']('/foobar')
  expect(vscode.window.showTextDocument).toBeCalledTimes(1)
})

test('_addSnippet', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  m['_gitProvider'] = { branch: 'foobar' } as any
  (m.getTextSelection as jest.Mock).mockReturnValueOnce({ text: '' })
  m['_addSnippet']({} as any)
  expect(vscode.window.showWarningMessage).toBeCalledWith('Marquee: no text selected');

  (m.getTextSelection as jest.Mock).mockReturnValueOnce({ text: 'foobar', name: 'some name', lang: 'ts' })
  m['_addSnippet']({ document: {
    uri: { path: path.sep + path.join('foo', 'bar') } }
  } as any)

  // delete path to make tests work in windows
  const calls = (m.updateState as jest.Mock).mock.calls
  delete calls[0][1][0].path
  expect(calls).toMatchSnapshot()
  expect((m['broadcast'] as jest.Mock).mock.calls).toMatchSnapshot()

  expect(vscode.commands.executeCommand).toBeCalledWith('marquee.refreshCodeActions')
  expect(vscode.window.showInformationMessage).toBeCalledWith(
    'Added some name to your snippets in Marquee',
    'Open Marquee'
  )
})

test('_insertFromTreeView', () => {
  delete vscode.window.activeTextEditor
  const m = new SnippetExtensionManager({} as any, {} as any)
  m['_insertFromTreeView']({ foo: 'bar' } as any)
  expect((vscode.window.showInformationMessage as jest.Mock).mock.calls[0][0]).toContain('No editor open')

  vscode.window.activeTextEditor = {
    edit: jest.fn(),
    revealRange: jest.fn(),
    document: 'foobar',
    selection: { active: { start: 123, end: 321 } }
  } as any
  m['_insertFromTreeView']({ isTreeItem: true, item: { body: 'bar' } } as any)
  expect(vscode.window.activeTextEditor?.edit).toBeCalledTimes(1)

  const cb = (vscode.window.activeTextEditor?.edit as jest.Mock).mock.calls[0][0]
  const editor = { insert: jest.fn() }
  cb(editor)

  expect(editor.insert).toBeCalledWith({ 'end': 321, 'start': 123 }, 'bar')
  expect(vscode.window.activeTextEditor?.revealRange)
    .toBeCalledWith({}, 'TextEditorRevealType.InCenterIfOutsideViewport')
  expect(vscode.window.showTextDocument).toBeCalledTimes(1)
})

test('_copyToClipboard', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  m['_copyToClipboard']({ item: { body: 'foobar' }} as any)
  expect(vscode.env.clipboard.writeText).toBeCalledWith('foobar')
})

test('_removeSnippet', () => {
  const m = new SnippetExtensionManager({} as any, {} as any)
  m.state.snippets = [{ id: 'foo' }, { id: 'bar' }, { id: 'loo' }] as any
  m['_removeSnippet']({ item: { id: 'bar' }} as any)
  expect((m.updateState as jest.Mock).mock.calls).toMatchSnapshot()
  expect((m['broadcast'] as jest.Mock).mock.calls).toMatchSnapshot()
})

test('activate', () => {
  const exp = activate({} as any, {} as any)
  expect(Object.keys(exp.marquee)).toMatchSnapshot()
})
