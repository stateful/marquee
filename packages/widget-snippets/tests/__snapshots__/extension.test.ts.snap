// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_addSnippet 1`] = `
Array [
  Array [
    "snippets",
    Array [
      Snippet {
        "archived": false,
        "body": "foobar",
        "createdAt": 1577836800000,
        "ctime": 1577836800000,
        "id": "123457890",
        "mtime": 1577836800000,
        "origin": undefined,
        "path": "/123457890/some name",
        "size": 6,
        "title": "some name",
        "type": "1",
        "workspaceId": null,
      },
      Object {
        "body": "bar",
        "id": "foo",
      },
    ],
  ],
]
`;

exports[`_addSnippet 2`] = `
Array [
  Array [
    Object {
      "snippets": Array [
        Snippet {
          "archived": false,
          "body": "foobar",
          "createdAt": 1577836800000,
          "ctime": 1577836800000,
          "id": "123457890",
          "mtime": 1577836800000,
          "origin": undefined,
          "path": "/123457890/some name",
          "size": 6,
          "title": "some name",
          "type": "1",
          "workspaceId": null,
        },
        Object {
          "body": "bar",
          "id": "foo",
        },
      ],
    },
  ],
]
`;

exports[`_removeSnippet 1`] = `
Array [
  Array [
    "snippets",
    Array [
      Object {
        "id": "foo",
      },
      Object {
        "id": "loo",
      },
    ],
  ],
]
`;

exports[`_removeSnippet 2`] = `
Array [
  Array [
    Object {
      "snippets": Array [
        Object {
          "id": "foo",
        },
        Object {
          "id": "loo",
        },
      ],
    },
  ],
]
`;

exports[`_saveNewSnippet 1`] = `
Array [
  Array [
    "snippets",
    Array [
      Object {
        "bar": "foo",
        "path": "foobar",
      },
      Object {
        "foo": "bar",
      },
    ],
  ],
]
`;

exports[`_saveNewSnippet 2`] = `
Array [
  Array [
    Object {
      "snippets": Array [
        Object {
          "bar": "foo",
          "path": "foobar",
        },
        Object {
          "foo": "bar",
        },
      ],
    },
  ],
]
`;

exports[`_updateSnippet 1`] = `
Array [
  Array [
    "snippets",
    Array [
      Object {
        "body": "bar",
        "id": "foo",
      },
    ],
  ],
]
`;

exports[`_updateSnippet 2`] = `
Array [
  Array [
    Object {
      "snippets": Array [
        Object {
          "body": "bar",
          "id": "foo",
        },
      ],
    },
  ],
]
`;

exports[`activate 1`] = `
Array [
  "disposable",
  "defaultState",
  "defaultConfiguration",
  "setup",
]
`;
